### PR TITLE
Adding single-device stress/perf test

### PR DESCRIPTION
**Pull request by CIPop**
- Updating stress test sln.
- Adding single device stress test.

### CODE REVIEW

**Comment by CIPop** [comment]
@prmathur-microsoft @abhipsaMisra  Note - this doesn't impact build at this time. 
1. We need to keep the build -stresstests up-to-date but that requires minor modifications of the tests to return non-zero on failure (which could be caused by high CPU, failing KPIs, etc).
2. We need to define a DevOps pipeline to run stress tests (these take considerably more and are highly dependent on network and VM SKU).

@jebrando - please let me know if we have a public common output spec for perf. We would like to adapt this as well as a future multi-device test to output the same CSV (or JSON).

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    public class DeviceTelemetryTest
+    {
+        int _messageSize;
+        int _messageCount;
+        int _parallelCount;
+        TransportType _protocol;
+        ResultWriter _writer;
+        DeviceClient _client;
```
Comment (davilu) [inline e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: `_client` could be local variable in `RunTestAsync`

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            _messageCount = messageCount;
+            _messageSize = messageSize;
+            _protocol = protocol;
+            _parallelCount = parallelCount;
+            _writer = resultWriter;
+
+            if (_messageCount < _parallelCount) throw new ArgumentException($"{nameof(messageCount)} needs to be greater than {nameof(parallelCount)}");
+        }
+
+        public void CreateDevice()
```
Comment (davilu) [inline e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: `private` and return `DeviceClient`

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            metrics.WallTime = _wall.Elapsed.TotalMilliseconds;
+
+            try
+            {
+                Stopwatch watch = new Stopwatch();
+                watch.Restart();
+                Message message = new Message(new byte[_messageSize]);
+                Task sendTask = _client.SendEventAsync(message);
+                watch.Stop();
+                metrics.SendMs = watch.Elapsed.TotalMilliseconds;
```
Comment (davilu) [inline e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: It's not `sendMs` but `task` was created. We're not able to get `sendMs`  in existing implementation.

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+                watch.Restart();
+                await sendTask.ConfigureAwait(false);
+                watch.Stop();
+                metrics.AckMs = watch.Elapsed.TotalMilliseconds;
```
Comment (davilu) [inline e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: This is time cost of sending and receiving ack.

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+
+            var sw = new Stopwatch();
+            sw.Start();
+            double startTimeSeconds = _wall.Elapsed.TotalSeconds;
+            int interimStatCompleted = 0;
+
+            for (int i = 0; i < _parallelCount; i++)
+            {
+                tasks.Add(SendMessageAsync());
```
Comment (davilu) [inline e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: In the existing stress test metrics, it took about 150 seconds to initialize 50K tasks. Start `_parallelCount` tasks here is not test `DeviceClient` performance but more like dotnet task scheduler.

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryMetrics.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            return
+                "WallTime," +
+                "Protocol, " +
+                "MessageSize," +
+                "SendMs, " +
+                "AckMs, " +
+                "ErrorMessage, ";
+        }
+
+        public override string ToString()
```
[the remark this answers is not in the document]
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryMetrics.cs]: The only concern would be the `null` check (not sure if that translates into "(null)" or something else).
In this case (no formatting) it appears that Join might be faster.

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    public class DeviceTelemetryTest
+    {
+        int _messageSize;
```
[the remark this answers is not in the document]
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: 👍

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs (hunk 1 above)
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: In C# is preferred to init in declaration. In dotnet/corefx we were following something similar to this: https://stackoverflow.com/questions/24551/initialize-class-fields-in-constructor-or-at-declaration (can't find the doc on dotnet/corefx).

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        int _messageCount;
+        int _parallelCount;
+        TransportType _protocol;
+        ResultWriter _writer;
+        DeviceClient _client;
+        Stopwatch _wall = new Stopwatch();
```
[the remark this answers is not in the document]
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: In C# is preferred to init in declaration. In dotnet/corefx we were following something similar to this: https://stackoverflow.com/questions/24551/initialize-class-fields-in-constructor-or-at-declaration (can't find the doc on dotnet/corefx).

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs (hunk 2 above)
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: 👍

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs (hunk 3 above)
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: SendMs measures the time (overheads) it takes to execute the `SendEventAsync` method.
Not sure what you mean by not being able to get sendMs (you mean until we write to the wire?) End-to-end send is actually measured as SendMs + AckMs.

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs (hunk 5 above)
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: I don't understand. Please clarify.
`_parallelCount` is one of the meta-parameters that need tuning based on the machine specification. In general it's small. I don't think we actually need 1000 here (I think ~100 will get the same result).

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs (hunk 1 above)
Comment (davilu) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]:
```
public async Task RunTestAsync()
{
    DeviceClient deviceClient = CreateDevice();
    ...
    await deviceClient.OpenAsync(timeout).ConfigAwait(false);
    ...
    for (int i = 0; i < _parallelCount; i++)
    {
        tasks.Add(SendMessageAsync(deviceClient));
    }
    ....
```

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryMetrics.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryMetrics.cs]: I take that back: these are double and int, etc - join only works on a string array.

File: e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs (hunk 1 above)
Comment (CIPop) [reply e2e/stress/DeviceTelemetryTest/DeviceTelemetryTest.cs]: I don't understand why I would need to pass in the deviceClient as a parameter to member functions when things are encapsulated in a single class.

**Comment by CIPop** [comment]
Closing for now: I have moved this into a more generic framework.

**Pull request closed without merging**